scenariorun/src/main/kotlin/com/cosmotech/scenariorun/azure/ScenarioRunServiceImpl.kt

```diff
@@ -71,309 +73,340 @@ internal class ScenarioRunServiceImpl(
+                scenarioRun.csmSimulationRun)
+
+            // It seems that deleteEntity does not throw any exception
+            logger.debug("Deleting Scenario Run {} from Cosmos DB", scenarioRun.id)
+            cosmosTemplate.deleteEntity("${scenarioRun.organizationId}_scenario_data", scenarioRun)
+            logger.debug("Scenario Run {} deleted from Cosmos DB", scenarioRun.id)
+        } catch (exception: IllegalStateException) {
+            logger.debug(
+                "An error occurred while deleting ScenarioRun {}: {}",
+                scenarioRun.id,
+                exception.message,
+                exception)
+        }
     }
-    deleteScenarioRunWithoutAccessEnforcement(scenarioRun)
-  }
-
-  private fun deleteScenarioRunWithoutAccessEnforcement(scenarioRun: ScenarioRun) {
-    // Simple way to ensure that we do not delete data if something went wrong
-    try {
-      logger.debug(
-          "Deleting scenario run. Organization: {}, Workspace: {}, Scenario Run Id: {}, csmSimulationRun: {}",
-          scenarioRun.organizationId ?: "null",
-          scenarioRun.workspaceKey ?: "null",
-          scenarioRun.id ?: "null",
-          scenarioRun.csmSimulationRun ?: "null")
-
-      azureDataExplorerClient.deleteDataFromADXbyExtentShard(
-          scenarioRun.organizationId!!, scenarioRun.workspaceKey!!, scenarioRun.csmSimulationRun!!)
-      logger.debug(
-          "Scenario run {} deleted from ADX with csmSimulationRun {}",
-          scenarioRun.id!!,
-          scenarioRun.csmSimulationRun)
-
-      // It seems that deleteEntity does not throw any exception
-      logger.debug("Deleting Scenario Run {} from Cosmos DB", scenarioRun.id)
-      cosmosTemplate.deleteEntity("${scenarioRun.organizationId}_scenario_data", scenarioRun)
-      logger.debug("Scenario Run {} deleted from Cosmos DB", scenarioRun.id)
-    } catch (exception: IllegalStateException) {
-      logger.debug(
-          "An error occurred while deleting ScenarioRun {}: {}",
-          scenarioRun.id,
-          exception.message,
-          exception)
+
+    override fun deleteObsoleteScenarioRunsByScenario(organizationId: String, workspaceId: String, scenarioId: String) {
+        val scenarioRuns = getScenarioRuns(organizationId, workspaceId, scenarioId)
+        val scenarioRunStatus = mutableListOf<ScenarioRunStatus>()
+        var successful = false
+        for(item in scenarioRuns){
+            scenarioRunStatus.add(getScenarioRunStatus(item.id!!, item))
+        }
+        scenarioRunStatus.sortedByDescending { it.endTime }
+        for(item in scenarioRunStatus){
+            val scenarioRunState = item.state
+            if(scenarioRunState == ScenarioRunState.Failed || scenarioRunState == ScenarioRunState.Unknown){
+                deleteScenarioRun(organizationId, item.id!!)
+            }else if(scenarioRunState == ScenarioRunState.Successful){
+                if(!successful){
+                    successful = true
```

Comment:
successsful not used here.
You should delete all successful scenario run which where launch prior to last successful run.
You can do that by removing it from scenarioRunStatus map or by ignoring the first successful run you find in your loop.
There is 1 addtional case: what about a scenario with no successful run: are we sure we want to delete all run.